rpcs3/util/serialization.hpp

```diff
@@ -0,0 +1,252 @@
+			//pos = -1;
+			//return false;
+		}
+
+		std::memcpy(ptr, &data[pos], size);
+		pos += size;
+		return true;
+	}
+
+	template <typename T>
+	bool read(T& obj)
+	{
+		return read(std::addressof(obj), sizeof(obj));
+	}
+
+	template <typename T>
+	bool read(T& obj) requires requires (T& obj) { obj.serialize(std::declval<serial_load&>()); }
+	{
+		obj.serialize(*this);
+		return operator bool();
+	}
+
+	// std::vector, std::basic_string
+	template <typename T>
+	bool read(T& obj) requires (std::is_constructible_v<typename T::value_type>) && requires (T& obj) { obj.data()[0]; obj.reserve(0); std::declval<serial_load&>().read(obj[0]); }
+	{
+		usz size = 0;
+
+		if (!read(size))
+		{
+			obj.clear();
+			return false;
+		}
+
+		obj.resize(size);
+
+		if constexpr (std::is_trivially_copyable_v<typename T::value_type>)
+		{
+			if (!read(obj.data(), sizeof(obj[0]) * size))
+			{
+				obj.clear();
+				return false;
+			}
+		}
+		else
+		{
+			for (auto&& value : obj)
+			{
+				if (!read(value))
+				{
+					obj.clear();
+					return false;
+				}
+			}
+		}
+
+		return true;
+	}
+
+	// std::array, C-array
+	template <typename T>
+	bool read(T& obj) requires (!requires (T& obj) { obj.clear(); }) && requires (T& obj) { std::size(obj); std::declval<serial_load&>().read(obj[0]); }
+	{
+		if constexpr (std::is_trivially_copyable_v<std::remove_reference_t<decltype(std::declval<T>()[0])>>)
+		{
+			return read(std::data(obj), sizeof(obj[0]) * std::size(obj));
+		}
+		else
+		{
+			for (auto&& value : obj)
+			{
+				if (!read(value))
+				{
+					return false;
+				}
+			}
+
+			return true;
+		}
+	}
+
+	// std::deque, std::set, std::map, std::multiset, std::multimap
+	template <typename T>
+	bool read(T& obj) requires (!requires (T& obj) { obj.data()[0]; }) && requires (T& obj) { obj.insert(obj.end(), std::declval<typename T::value_type>()); }
+	{
+		usz size = 0;
+
+		if (!read(size))
+		{
+			obj.clear();
+			return false;
+		}
+
+		for (usz i = 0; i < size; i++)
+		{
+			if constexpr (std::is_trivially_copyable_v<typename T::value_type>)
+			{
+				char buf[sizeof(typename T::value_type)]{};
+
+				if (!read(buf))
+				{
+					obj.clear();
+					return false;
+				}
+
+				obj.insert(obj.end(), std::bit_cast<typename T::value_type>(buf));
+			}
+			else
+			{
+				typename T::value_type value{};
+
+				if (!read(value))
+				{
+					obj.clear();
+					return false;
+				}
+
+				obj.insert(obj.end(), std::move(value));
+			}
+		}
+
+		return true;
+	}
+
+	template <typename T> requires (std::is_copy_constructible_v<T> && (std::is_constructible_v<T> || std::is_trivially_copyable_v<T>))
+	operator T()
+	{
+		if constexpr (std::is_constructible_v<T>)
+		{
+			T value{};
+			read(value);
+			return value;
+		}
+		else
+		{
+			char buf[sizeof(T)]{};
+			read(buf);
+			return std::bit_cast<T>(buf);
+		}
+	}
+
+	explicit operator bool() const
+	{
+		return pos <= data.size();
+	}
+
+	template <typename... T>
+	bool operator()(T&... args)
```

Comment:
I think using this operator is not very intuitive.
Maybe use << and >> respectively

Reply:
Or set and get

Reply:
Most serialization libraries do it this way and it is very convenient. Furthermore I would like to not change existing code for cereal in RSX captures.